File: src/Belajar_SOLID_14_Dependency_Inversion_Principle/Java/Main.java
package Belajar_SOLID_14_Dependency_Inversion_Principle.Java;

class PaymentService{
    private Database database;

    public PaymentService(Database database){
        this.database = database;
    }

    void paymentIsValid(){}
    void openDatabase(){}
    void addNewPayment(){}
    void removePaymentById(){}
    void updatePaymentById(){}
}

abstract class Database{
    abstract void insert();
    abstract void update();
    abstract void delete();

}

class MySQLDatabase extends Database{
    @Override
    void insert() {

    }

    @Override
    void update() {

    }

    @Override
    void delete() {

    }
}

class MongoDatabase extends Database{

    @Override
    void insert() {

    }

    @Override
    void update() {

    }

    @Override
    void delete() {

    }
}

File: src/Belajar_SOLID_14_Dependency_Inversion_Principle/Kotlin/Main.kt
package Belajar_SOLID_14_Dependency_Inversion_Principle.Kotlin

import javax.xml.crypto.Data

class PaymentService(val database:Database){

    fun paymentIsValid(){

    }

    fun openDatabase(){

    }

    fun addNewPayment(){

    }

    fun removePaymentById(){

    }

    fun updatePaymentById(){

    }
}

abstract class Database{
    abstract fun insert()
    abstract fun update()
    abstract fun delete()
}

class MySQLDatabase : Database(){
    override fun insert() {
        TODO("Not yet implemented")
    }

    override fun update() {
        TODO("Not yet implemented")
    }

    override fun delete() {
        TODO("Not yet implemented")
    }
}

class MongoDatabase : Database(){
    override fun insert() {
        TODO("Not yet implemented")
    }

    override fun update() {
        TODO("Not yet implemented")
    }

    override fun delete() {
        TODO("Not yet implemented")
    }

}